### PR TITLE
Feature/#82 get chats from firestore

### PR DESCRIPTION
**Pull request by roniash101**

### CODE REVIEW

File: client/src/components/ChatItem.tsx [hunk 1]
```diff
@@ -1,35 +1,114 @@
...
-  name: string;
-  isEnded: boolean;
-  chatId: string;
-}
+  name: string,
+  isEnded: boolean,
+  chatId: string
+};
 
 export const ChatItem: React.FC<ChatItemProps> = ({ name, isEnded, chatId }: ChatItemProps) => {
```
Comment (RealByter) [inline client/src/components/ChatItem.tsx]: React.FC<ChatItemProps> already defines the type for the props so there's no need to define the type for the props separately (idk if it can cause issues)

File: client/src/components/ChatItem.tsx [hunk 2]
```diff
@@ -1,35 +1,114 @@
...
+import { getRealtimeLastMessage, getUnreadMessagesCount } from '../helpers/chatFunctions';
+import { Message } from '../firebase/message';
+import { getCurrDateIsrael, isYesterday } from '../helpers/dateFunctions';
 
 export interface ChatItemProps {
-  name: string;
-  isEnded: boolean;
-  chatId: string;
-}
+  name: string,
```
Comment (RealByter) [inline client/src/components/ChatItem.tsx:11]: Consider changing this field name to supporterName for easier readability. I for once got confused with our user's name

File: client/src/components/Message.tsx [hunk 3]
```diff
@@ -1,18 +1,26 @@
...
+export const Message: React.FC<IMessageProps> = ({ messageId, isSender, content, messageDate, messageState }: IMessageProps) => {
+
+  useEffect(() => {
+    if (isSender) return;
+    if (!messageId) return;
+
+    (async () => {
+      await updateDoc(doc(collections.messages, messageId), { status: "read" });
+    })();
+  }, [messageId])
```
Comment (RealByter) [inline client/src/components/Message.tsx]: There should also be a check if the messagesState is already 'read' to prevent unnecessary calls to firebase

File: client/src/components/Message.tsx [hunk 4]
```diff
@@ -1,18 +1,26 @@
...
-  isSender,
-  content,
-  messageDate,
-  messageState
-}: IMessageProps) => {
+export const Message: React.FC<IMessageProps> = ({ messageId, isSender, content, messageDate, messageState }: IMessageProps) => {
+
+  useEffect(() => {
+    if (isSender) return;
+    if (!messageId) return;
```
Comment (RealByter) [inline client/src/components/Message.tsx]: Can the Message component even exist without a messageId?

File: client/src/helpers/chatFunctions.tsx [hunk 5]
```diff
@@ -8,17 +8,73 @@ import {
...
+    where(roleFieldName, '==', userId),
+    orderBy('createdAt')
+  );
+
+  const unsubscribe = onSnapshot(queryUserChats, (snapshot) => {
+    const queryData = snapshot.docs.map((doc) => doc.data());
+    const filteredData = queryData.filter((doc) => doc[oppositeRoleFieldName] !== userId && doc.status != "blocked");
+    cb(filteredData);
+  }, OnSnapshotError);
+  return unsubscribe;
```
Comment (RealByter) [inline client/src/helpers/chatFunctions.tsx]: The filtering here should be in the query itself. Right now you're reading a lot more documents than you even need. I'm talking about the checks to see if the userId is also the opposite role and if the status is not blocked

File: client/src/components/ChatItem.tsx [hunk 6]
```diff
@@ -1,35 +1,114 @@
...
+    {
+      return otherTime;
+    }
+    else if (isYesterday(otherDate, currDate)) {
+      return "אתמול";
+    }
+    else {
+      return otherDate.replace("/", ".").replace("/", "."); // twice in orderto replace all
+    }
+  }
```
Comment (RealByter) [inline client/src/components/ChatItem.tsx]: If you already have the dateFunctions.ts for functions that are related to dates maybe you can clean this file a bit by moving this function to that file

File: client/src/pages/Chat.tsx [hunk 7]
```diff
@@ -54,7 +55,7 @@ export const Chat = () => {
     if (!chatLoading && !chat) navigate('/selection');
   }, [chat, chatLoading, navigate]);
 
-  useEffect(() => {
+  useEffect(() => { // todo: getName from chatObj instead?
```
Comment (RealByter) [inline client/src/pages/Chat.tsx]: What is this chatObj?

File: client/src/pages/ChatsListPage.tsx [hunk 8]
```diff
@@ -1,80 +1,87 @@
...
-          />
-        ))}
-        {endedChats && endedChats.length > 0 ? (
-          <>
-            <p className="chats-ended">שיחות שהסתיימו</p>
-            {endedChats?.map((chat) => (
+  useEffect(() => {
+    if (!user) return;
+
+    // todo: make more generic
```
Comment (RealByter) [inline client/src/pages/ChatsListPage.tsx:28]: Not sure what you meant here

File: client/src/styles/components/_chats-list.scss [hunk 9]
```diff
@@ -1,103 +1,148 @@
...
-    font-weight: 500;
-    margin-bottom: 1.75rem;
-    font-family: $font-secondary;
-  }
-
-  .list {
+  &>.content {
+    position: fixed;
+    width: 100vw;
+    height: 65vh;
```
Comment (RealByter) [inline client/src/styles/components/_chats-list.scss]: This is a very bad practice that I've seen in the code before but didn't know who wrote it. You should almost never use specific vh/vw units for 2 main reasons:

1. It's not necessarily accurate to the figma design
2. When the screen is changing it might not always scale how we'd expect. For example - earlier in the project someone made the send icon be relative to the screen width. As you can see in the figma design it's always a circle. So the actual result was that it shrinked/widened incorrectly when it shouldn't even change size

Try to almost always use flexbox, margins and paddings to make it accurate to the design in figma. If you didn't know, you can press the toggle on the top right to switch to 'developer mode'. This allows you to click on any element in figma to view the paddings and margins in pixels. Saves a lot of time, calculations and effort

File: client/src/styles/components/_chats-list.scss [hunk 10]
```diff
@@ -1,103 +1,148 @@
...
+
+  .loading {
+    color: $default-text-color;
+    margin-right: 2rem;
+  }
+
+  .chat-item {
+    background: rgba($color: white, $alpha: 0.8);
+    padding: 0.75rem;
+    margin: 0 9vw 1rem 9vw;
```
Comment (RealByter) [inline client/src/styles/components/_chats-list.scss]: Same as before

File: client/src/styles/components/_chats-list.scss [hunk 11]
```diff
@@ -1,103 +1,148 @@
...
-  .new-supportee {
-    margin: auto 0 1.25rem;
-    padding-top: 1.75rem;
+  .footer {
+    box-shadow: 0px -20px 35px $dark-purple;
+    background-color: $dark-purple;
+    position: fixed;
+    bottom: 0;
+    width: 100vw;
+    height: 18vh;
```
Comment (RealByter) [inline client/src/styles/components/_chats-list.scss]: Same as before. Just making sure that it's understandable that I'm talking about the specific values and not the 100vw/vh - which is okay

File: client/src/styles/components/_chats-list.scss [hunk 12]
```diff
@@ -1,103 +1,148 @@
...
+    display: flex;
+    flex-direction: column;
+    align-items: center;
+    justify-content: space-evenly;
 
-    button {
+    .new-supportee {
+      background-color: white;
+      color: $default-text-color;
+      width: 83vw;
```
Comment (RealByter) [inline client/src/styles/components/_chats-list.scss]: Same as before

**Review by RealByter — CHANGES_REQUESTED** [review]
Overall great code. I liked how you handled the functionality of marking messages as 'read', much more efficient than I thought of myself

File: client/src/components/ChatItem.tsx (hunk 1 above)
Comment (roniash101) [reply client/src/components/ChatItem.tsx]: Hi,
how would you expect it to be written otherwise? do you have an example that I could use as a reference?

File: client/src/components/ChatItem.tsx (hunk 2 above)
Comment (roniash101) [reply client/src/components/ChatItem.tsx:11]: I am avoiding calling it something more specific in case we'd at some point want to use that component for the other role as well,
and I actually think that from the small "component's point of view" it's pretty clear :)

File: client/src/components/Message.tsx (hunk 4 above)
Comment (roniash101) [reply client/src/components/Message.tsx]: Probably not, fixed.

File: client/src/helpers/chatFunctions.tsx (hunk 5 above)
Comment (roniash101) [reply client/src/helpers/chatFunctions.tsx]: The problem is that firebase doesn't allow to write a query that uses

- OrderBy(key1) and where(key2, '!=', value)

or even 

- where(key1, '!=', value) and where(key2, '!=', value)

(In firebase docs: "In a compound query, range (<, <=, >, >=) and not equals (!=, not-in) comparisons must all filter on the same field."

It means that it is either ordering by the creation date and filtering out the unneeded results in the client side
or
filtering out some/most of the unwanted results in the query but also sorting all the results according to their creation date in the client side

Bottom line, we need to decide which one is preferable.

File: client/src/pages/ChatsListPage.tsx (hunk 8 above)
Comment (roniash101) [reply client/src/pages/ChatsListPage.tsx:28]: In line 28:
`    let role: Role = "supporter";
`
For future perposes, if we want the "supportee" to have a chatsList page as well, the `role` variable should be generic instead of hard-coded.

File: client/src/styles/components/_chats-list.scss (hunk 9 above)
Comment (roniash101) [reply client/src/styles/components/_chats-list.scss]: Interesting, I actually didn't know about the developer mode.
Also, from what I heard using px is not a good practice and rem is preferable.
Anyway, I checked in the InspectTools and the ChatsListPage is responsive to all devices except Galaxy Fold, so I am not too convinced it needs to be changed... what do you think?

File: client/src/components/ChatItem.tsx (hunk 1 above)
Comment (RealByter) [reply client/src/components/ChatItem.tsx]: Just remove the `: ChatItemProps` after the props deconstruction

File: client/src/helpers/chatFunctions.tsx (hunk 5 above)
Comment (RealByter) [reply client/src/helpers/chatFunctions.tsx]: I looked at this again and I'm not even sure why the 2 other conditions are there.. We have no blocking functionality by now and there is no way that the oppositeRoleFieldName would ever be the same as the roleFieldName

File: client/src/styles/components/_chats-list.scss (hunk 9 above)
Comment (RealByter) [reply client/src/styles/components/_chats-list.scss]: Rem and px is controversial most of the time. font sizes 100% needs to be with rem, it's the margins and paddings that are not always like that. I personally always use rem just because I like it. I still think it needs to be changed to respond to the layout rather than "fixating on itself" with margins and paddings in addition to the first reason. It's just hard to predict where it would fail basically. Also when something changes in the design, most of the time this is still going to be styled correctly with margins and paddings rather than with specific sizes

File: client/src/components/Message.tsx (hunk 3 above)
Comment (roniash101) [reply client/src/components/Message.tsx]: good call!

File: client/.evv
Comment (osherh) [inline client/.evv:1]: this file is empty
i think it is redundant

File: package.json [hunk 13]
```diff
@@ -0,0 +1 @@
+{}
```
Comment (osherh) [inline package.json:1]: package.json doesnt have react lib even though u use react

File: client/src/styles/setup/_variables.scss [hunk 14]
```diff
@@ -7,7 +7,7 @@ $chat-first-part-msg-bgd: #ffffff;
 $chat-second-part-msg-bgd: #9360ff;
 $chat-send-btn-bgd: #e5dbff;
 $scrollbar-bgd: #e5dbff;
-$dark-purple: #baa6f1; // todo: remove
+$dark-purple: #baa6f1;
```
Comment (osherh) [inline client/src/styles/setup/_variables.scss]: give meaningful color names
$primary-color: #e5dbff
such that if u change the color u will change it once as u use the name primary-color on the app

File: client/src/helpers/chatFunctions.tsx (hunk 5 above)
Comment (roniash101) [reply client/src/helpers/chatFunctions.tsx]: "We have no blocking functionality by now" -> We are obviously going to have the blocking functionality so it is crucial to support it...
"There is no way that the oppositeRoleFieldName would ever be the same as the roleFieldName" -> You are right, i'll remove it.

File: client/src/pages/Chat.tsx (hunk 7 above)
Comment (roniash101) [reply client/src/pages/Chat.tsx]: Regarding to line 62:
`      const compName = await getNameById(chat![getOppositeRoleFieldName(userRole)]!);
`
maybe we should add another field to the chatSchema and then use that instead of requesting it from the server every time

const chatSchema = z.object({
  id: z.string(),
  supporterId: z.string().nullable(),
  supporteeId: z.string().nullable(),
  createdAt: z.instanceof(Timestamp),
  status: statusEnum.default('active'),
  supporteeName: z.string().nullable(),
**supporterName: z.string().nullable()**
});

File: client/.evv
Comment (roniash101) [reply client/.evv:1]: It is needed in order to activate the firestore simulator :)

File: package.json (hunk 13 above)
Comment (roniash101) [reply package.json:1]: This file is created automatically so I don't really know what happend, thank you for pointing it out

**Review by RealByter — APPROVED** [review]
